### PR TITLE
Implement DATEDIF function

### PR DESCRIPTION
**Pull request by MiNGRotten**
Implement DATEDIF function for calculate difference between two dates.
MS specs: https://support.microsoft.com/en-us/office/datedif-function-25dba1a4-2812-480b-84dd-8b32a451b35c

### CODE REVIEW

**Comment by igitur** [comment]
I refactored the method. There were some edge cases that weren't calculated correctly.

File: ClosedXML/Excel/CalcEngine/Functions/DateAndTime.cs [hunk 1]
```diff
@@ -113,6 +115,31 @@ private static object Date(List<Expression> p)
...
+            DateTime endDate = p[1];
+            string unit = p[2];
+
+            if (startDate > endDate)
+                throw new NumberException("The start date is greater than the end date");
+
+            return (unit.ToUpper()) switch
+            {
+                "Y" => endDate.Year - startDate.Year - (new DateTime(startDate.Year, endDate.Month, endDate.Day) < startDate ? 1 : 0),
+                "M" => Math.Truncate((endDate.Year - startDate.Year) * 12d + endDate.Month - startDate.Month),
```
Comment (Pankraty) [inline ClosedXML/Excel/CalcEngine/Functions/DateAndTime.cs:130]: Nope.
image

**Review by Pankraty — CHANGES_REQUESTED** [review]
Not all edge cases have been covered, apparently

File: ClosedXML/Excel/CalcEngine/Functions/DateAndTime.cs (hunk 1 above)
Comment (igitur) [reply ClosedXML/Excel/CalcEngine/Functions/DateAndTime.cs]: Good catch. Fixed.

**Comment by igitur** [comment]
I'll squash and rebase after final approval.

File: ClosedXML/Excel/CalcEngine/Functions/DateAndTime.cs [hunk 2]
```diff
@@ -113,6 +115,31 @@ private static object Date(List<Expression> p)
...
+            string unit = p[2];
+
+            if (startDate > endDate)
+                throw new NumberException("The start date is greater than the end date");
+
+            return (unit.ToUpper()) switch
+            {
+                "Y" => endDate.Year - startDate.Year - (new DateTime(startDate.Year, endDate.Month, endDate.Day) < startDate ? 1 : 0),
+                "M" => Math.Truncate((endDate.Year - startDate.Year) * 12d + endDate.Month - startDate.Month - (endDate.Day < startDate.Day ? 1 : 0)),
+                "D" => Math.Truncate(endDate.Subtract(startDate).TotalDays),
```
Comment (Pankraty) [inline ClosedXML/Excel/CalcEngine/Functions/DateAndTime.cs:131]: Nope. 
image

This should do the trick:
```suggestion
                "D" => Math.Truncate(endDate.Date.Subtract(startDate.Date).TotalDays),
```

Or even without `Math.Truncate` - CLR does not seem to care about daylight saving time and returns whole numbers anyway.

**Comment by igitur** [comment]
Now I understand why Microsoft discourages use of this function ;-)

**Pull request merged**